### PR TITLE
Add support to set custom HTTP Status codes

### LINKED ISSUE #39: Use HTTP Integer codes as the argument to createResponseBuilder() method NOT enums
I have been told this is the place to raise this (again - apologies for listing in multiple places)
Until 2.0.11933, (almost 1 YEAR), the createResponse() method signature has used an integer HTTP code. We have built functions that rely on returning non-standard (e.g. 460) Http codes for INTERNAL consumption (within out framework) only. The new method (createResponseBuilder) currently requires an enum that represents a "standard" HTTP code. It is painful enough to have to change source code to use a different method (createResponseBuilder rather than createResponse), but we MUST be able to generate Http codes rather than relying on those that MS chooses to define as an enum. While I generally agree with using enums as parameters, in this case since the enum MUST be converted back to integer for HTTP protocols, there is no compelling reason for MS to interject an enum mapping - which also has the side effect of prohibiting custom codes. If you must, put a sanity check for integers below 100 or above 1000, but please do not prohibit custom codes. I note also that anyone else who is seriously using Java Http trigger functions by necessity would have been using integers - so I don't think making the new method signature revert to integers is an issue for current developers - everyone's source code is going to have to change just to accommodate the new method.
 Thanks.
 TurtleBeach AKA -jack-

### PR DESCRIPTION
**Pull request by brunoborges**
Fixes #39

Signed-off-by: Bruno Borges <bruno.borges@microsoft.com>

### CODE REVIEW

File: src/main/java/com/microsoft/azure/functions/HttpResponseMessage.java [hunk 1]
```diff
@@ -22,7 +22,7 @@
      * 
      * @return the status code set on the HttpResponseMessage instance.
      */
-    HttpStatus getStatus();
+    int getHttpStatus();
```
Comment (pragnagopa) [inline src/main/java/com/microsoft/azure/functions/HttpResponseMessage.java]: Can we have both? Giving user a choice between using enums vs int

**Comment by pragnagopa** [comment]
Can we have both? Giving user a choice between using enums vs int

**Comment by brunoborges** [comment]
We can't. HttpStatus enum is for standard http codes. If user sets a custom code, getStatus would have to return null, which is error prone in user code.

File: src/main/java/com/microsoft/azure/functions/HttpResponseMessage.java (hunk 1 above)
Comment (brunoborges) [reply src/main/java/com/microsoft/azure/functions/HttpResponseMessage.java]: We can't. HttpStatus enum is for standard http codes. If user sets a custom code, getStatus would have to return null, which is error prone in user code.

**Comment by pragnagopa** [comment]
cc: @paulbatum 
I have looked into few libraries. Having both enum status and int statusCode seems to be very common
- https://docs.oracle.com/javaee/6/api/javax/ws/rs/core/Response.Status.html
- https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/http/HttpStatus.html
And for the srpingframework, the documentation does state 

```
static HttpStatus | resolve(int statusCode)Resolve the given status code to an HttpStatus, if possible.
```
Original PR: https://github.com/Azure/azure-functions-java-library/pull/12 added the enums to address issue https://github.com/Azure/azure-functions-java-worker/issues/74, which is still true.

**Please note that replacing the API to  return int is a breaking change**.  At this point, adding support to return custom codes needs to be considered as a feature improvement and not breaking the api changes we just made and rolled out.

**Comment by JonathanGiles** [comment]
One option to explore is to have the enum implement an interface that had the relevant API on it, and then support custom codes in a separate class or as a default method on the interface. This interface becomes the return types rather than the enum class.

File: src/main/java/com/microsoft/azure/functions/HttpResponseMessage.java (hunk 1 above)
Comment (JunyiYi) [reply src/main/java/com/microsoft/azure/functions/HttpResponseMessage.java]: Why not throw a runtime exception when the custom code is not in the standard HTTP codes?

**Comment by JunyiYi** [comment]
@JonathanGiles , just my opinion, although quite elegant in large products, but your suggestion will make this simple interface more complicated.

**Comment by JunyiYi** [comment]
@brunoborges Be careful when introducing incompatible changes. We need to update all documentations and templates all over the place when you change this interface.

**Comment by brunoborges** [comment]
@JunyiYi I understand the concern, but given this interface is to be consumed only (read), we should make an effort to get the design right. JAX-RS design, as pointed by @pragnagopa, is the most accurate to what we want (standard and custom http status codes).

I can spend some cycles here between today and Thursday to get something done.

**Comment by pragnagopa** [comment]
@brunoborges  - Any update on this?

**Comment by pragnagopa** [comment]
Thanks! Will review.

File: src/main/java/com/microsoft/azure/functions/HttpStatusType.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * Defines an HTTP Status Type
+ * 
+ * @since 1.0
+ */
+public interface HttpStatusType {
+
+    public static enum StatusClass {
+        INFORMATIONAL(1), SUCCESS(2), REDIRECTION(3), CLIENT_ERROR(4), SERVER_ERROR(5), OTHER(0);
```
Comment (pragnagopa) [inline src/main/java/com/microsoft/azure/functions/HttpStatusType.java]: What is the purpose of statusclass? Can you please give an example?

File: src/main/java/com/microsoft/azure/functions/HttpStatusType.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
+package com.microsoft.azure.functions;
+
+/**
+ * Defines an HTTP Status Type
+ * 
+ * @since 1.0
+ */
+public interface HttpStatusType {
```
Comment (pragnagopa) [inline src/main/java/com/microsoft/azure/functions/HttpStatusType.java]: Move this to a its own file

File: src/main/java/com/microsoft/azure/functions/HttpStatusType.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    public static enum StatusClass {
+        INFORMATIONAL(1), SUCCESS(2), REDIRECTION(3), CLIENT_ERROR(4), SERVER_ERROR(5), OTHER(0);
+        int value = 0;
+
+        StatusClass(int _class) {
+            this.value = _class;
+        }
+
+        public static StatusClass statusClassOf(int code) {
+            int computed = code / 100;
```
Comment (pragnagopa) [inline src/main/java/com/microsoft/azure/functions/HttpStatusType.java]: Add validation here if (code <= 0)  as well

File: azure-pipelines.yml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+# Add steps that analyze code, save build artifacts, deploy, and more:
+# https://docs.microsoft.com/vsts/pipelines/languages/java
+
+pool:
+  vmImage: 'Ubuntu 16.04'
+
+steps:
+- task: Maven@3
+  inputs:
+    mavenPomFile: 'pom.xml'
```
Comment (pragnagopa) [inline azure-pipelines.yml]: Is this used anywhere?

File: src/main/java/com/microsoft/azure/functions/HttpStatus.java [hunk 6]
```diff
@@ -39,22 +42,39 @@
     BANDWIDTH_LIMIT_EXCEEDED(509), NOT_EXTENDED(510), NETWORK_AUTHENTICATION_REQUIRED(511);
 
     private final int value;
+    private final String reason;
 
     HttpStatus(int value) {
         this.value = value;
+        this.reason = normalizeString(this.name());
```
Comment (pragnagopa) [inline src/main/java/com/microsoft/azure/functions/HttpStatus.java]: Please add unit tests.

**Comment by pragnagopa** [comment]
@brunoborges - Apologies for the delayed review on this. We are tracking this for GA. Changes mostly look good. Added few comments/questions

If you do not have cycles to address feedback and pull it through to azure-functions-java-worker, I will take this over.

**Comment by brunoborges** [comment]
Hi @pragnagopa, I am traveling this week so yes please, take it from here.

File: src/main/java/com/microsoft/azure/functions/HttpStatusType.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                }
+            }
+
+            return OTHER;
+        }
+    }
+
+    public int getCode();
+
+    public String getReason();
```
Comment (pragnagopa) [inline src/main/java/com/microsoft/azure/functions/HttpStatusType.java]: Remove getReason. 
Setting ReasonPhrase is hidden in AspNet ObjectResult https://github.com/aspnet/HttpAbstractions/issues/395. Will add support if there is an ask

**Comment by pragnagopa** [comment]
Closing this PR. Will send a new one.

**Pull request closed without merging**